controllers/volumereplicationgroup_controller.go

```diff
@@ -1055,7 +985,32 @@ func (v *VRGInstance) relocate(result *ctrl.Result, s3StoreAccessors []s3StoreAc
 	}
 }
 
-func (v *VRGInstance) updateVRGStatus(updateConditions bool) error {
+func (v *VRGInstance) invalid(err error, msg string, requeue bool) ctrl.Result {
+	rmnutil.ReportIfNotPresent(v.reconciler.eventRecorder, v.instance, corev1.EventTypeWarning,
+		rmnutil.EventReasonValidationFailed, err.Error())
+
+	return v.dataError(err, msg, requeue)
+}
+
+func (v *VRGInstance) dataError(err error, msg string, requeue bool) ctrl.Result {
+	return v.err(err, msg, requeue, setVRGDataErrorCondition)
+}
+
+func (v *VRGInstance) clusterDataError(err error, msg string, requeue bool) ctrl.Result {
+	return v.err(err, msg, requeue, setVRGClusterDataErrorCondition)
+}
+
+func (v *VRGInstance) err(err error, msg string, requeue bool,
+	conditionSet func(*[]metav1.Condition, int64, string),
+) ctrl.Result {
+	v.log.Info(msg, "error", err)
+	msg = fmt.Sprintf("%s: %v", msg, err)
```

Comment:
Shadowing msg is not nice and make it harder to follow the code.

Unifying the error path so we always use the one of the *Error functions looks good but sending a function for updating the condition is little bit too clever and the benefit is very small.

Since we always log the msg and err and set the condition with the msg and error (formatted) we can move the log and formatting down the condition set function.

    func (v *VRGInstance) dataError(err error, msg string, requeue bool) ctrl.Result {
        setVRGDataErrorCondition(&v.instance.Status.Conditions, v.instance.Generation, msg, err)
	    return v.updateVRGStatus(ctrl.Result{Requeue: requeue}, false)
    }
    
    func (v *VRGInstance) clusterDataError(err error, msg string, requeue bool) ctrl.Result {
	    setVRGClusterDataErrorCondition(&v.instance.Status.Conditions, v.instance.Generation, msg, err)
        return v.updateVRGStatus(ctrl.Result{Requeue: requeue}, false)
    }

Reply:
OK